### DISCUSSION TITLE
Cannot compile `mdxjsEsm` node

### DISCUSSION THREAD

**Discussion by lancejpollard** [mdx-js/mdx#2224] [Q&A]
I am in Next.js with this config:

```js
/** @type {import('next').NextConfig} */
import createMDXPlugin from '@next/mdx'
import rehypeRaw from 'rehype-raw'

const mdxConfig = {
  extension: /\.mdx?$/,
  options: {
    providerImportSource: '@mdx-js/react',
    rehypePlugins: [rehypeRaw],
    // If you use remark-gfm, you'll need to use next.config.mjs
    // as the package is ESM only
    // https://github.com/remarkjs/remark-gfm#install
    remarkPlugins: [],
  },
}

const withMDX = createMDXPlugin(mdxConfig)

const nextConfig = withMDX({
  compiler: {
    styledComponents: true,
  },
  pageExtensions: ['mdx', 'md', 'tsx', 'ts'],
  poweredByHeader: false,
  // reactStrictMode: true,
  trailingSlash: false,
})

export default nextConfig
```

I have installed the 3 modules listed [here](https://nextjs.org/docs/advanced-features/using-mdx). I can successfully render some MDX content which I import into a next.js page. But when I try to load a `pages/example.mdx` file directly as a Next.js page, it throws this error:

```
error - ./pages/cellular-automata/index.mdx
Error: index.mdx:Error: Cannot compile `mdxjsEsm` node
    at new Promise (<anonymous>)
```

Searching the web for `mdxjsEsm` yields nothing, surprisingly.

All I have in my mdx file is this:

```mdx
import Layout from '@components/kit/Layout'

export const meta = {
  url: '/cellular-automata',
  tab: 'Cellular Automata Notes @ Lance Pollard',
  keywords: ['cellular automata'],
}

export default ({ children }) => <Layout {...meta}>{children}</Layout>
```

Any ideas? Layout renders the Next.js head `import NextHead from 'next/head'`, so maybe that is causing issues? I don't know, could use some help.

**Comment by lancejpollard** [comment]
If I change `next.config.js` to `next.config.mjs`, it now throws a different error:

```
Error: poem.md:Error: Cannot compile `mdxJsxFlowElement` node
    at new Promise (<anonymous>)
Import trace for requested module:
./contents/poem.md
```

**Reply by icyJoseph** [reply]
It's just how it is transforming the mdx file into JavaScript, either with ESM or CJS, but the underlying error is the same. I think anyway.

**Comment by icyJoseph** [comment]
Could you put some text content in the mdx file and try again?

**Reply by lancejpollard** [reply]
Same error if I have this:

```
import Layout from '@components/kit/Layout'

export const meta = {
  url: '/cellular-automata',
  tab: 'Cellular Automata Notes @ Lance Pollard',
  keywords: ['cellular automata'],
}

I am text.

export default ({ children }) => <Layout {...meta}>{children}</Layout>
```

**Comment by wooorm** [comment]
Remove `rehype-raw`, `rehype-raw` has quite specific complex use cases, you don’t indicate that you need it.

**Reply by lancejpollard** [reply]
I need to be able to include native HTML elements in the markdown, how can I do that without rehype-raw? Here are two files I have so far:

poem.md:

```md
<ul>
  <li>The moon observes from its perch, high above the peak,</li>
  <li>Telling other creatures, their task is never complete.</li>
  <li>Far beneath the mountain a valley is carved,</li>
  <li>As a river churns for endless years, through stone so hard.</li>
</ul>
<ul>
  <li>Approaching like a hunter, the infinite byte</li>
  <li>Lighting the way for others to build into the night,</li>
  <li>The valley, the breeze, the animals, the trees,</li>
  <li>All working together, opening our eyes to see.</li>
</ul>
<ul>
  <li>Basking in the warm sun and breathing the mountain air.</li>
  <li>Learning from so many, working together is all that's fair.</li>
  <li>All that we call nature is constantly changing.</li>
  <li>Each day, each hour, every minute rearranging.</li>
</ul>
<ul>
  <li>New paths taken as friends come along,</li>
  <li>Learning the outcomes in nature are rarely wrong.</li>
  <li>There to advise and to learn,</li>
  <li>Making sure progress survives every turn.</li>
</ul>
```

description.md

```
My ears twitch, nose wrinkles, and dawn is approaching. The birds are
chirping, singing, a choir of melodies. I stretch, yawn, stand and greet
the sun. No hint of rain or snow as I kick, cautiously thumping over the
trodden ground of my home. It is time to feed, nibbling on the long
grass dancing in the pink morning dew. It is a day for work; my quest
will not complete itself. There is more to discover here - deeper
caverns to dig - and I know that within them lies the truth. It is here,
close at hand; the harder I work, the greater the excitement. I am so
close, creating and discovering the truth of it all. A yearning for
knowledge, a curiosity unsated as the day draws to a close. Night is
coming and I know I must stop. Regret pulls at me from within and I feel
it slow. The sky is orange, the stars are coming soon as the flowers
have closed their petalled eyes. I am satisfied, appreciating the
accomplishment before me. It is time to sleep and nestle in for the
night, listening to the rising sounds of the crickets and frogs, a
symphony to lure me into dreams. A job well done, another day gone, and
visions fall upon me of endless greens and wide azure skies.
```

That latest file is the one with code in it (first post).

**Reply by wooorm** [reply]
Why not use MDX? Sure, you *could* have some plain markdown files and used them with our loaders.
But that’s a good idea when you’re migrating tons and tons of content.
It’s a lot better to use MDX with MDX

**Reply by lancejpollard** [reply]
What are you suggesting then? I don't follow.

**Reply by lancejpollard** [reply]
I'm getting quite frustrated with MDX, it's taken hours to get a hello-world example working in Next.js, and now it seems like there are more problems (https://github.com/vercel/next.js/discussions/44883), but first I'd like to see us get past this current one which we've been discussing here.

**Reply by lancejpollard** [reply]
When I wasn't using `rehype-raw`, the poem (which has HTML in it) rendered empty. Any ideas on that then?

**Reply by wooorm** [reply]
To use an `.mdx` extension and to use the MDX format: https://mdxjs.com/docs/what-is-mdx/#mdx-syntax

**Reply by wooorm** [reply]
> I'm getting quite frustrated with MDX

Sad that you are having trouble. But this is open source. You get this code for free, whether it works or not. You *even* get people that try to help you, but the dynamics with that are that there are 100s of people that ask questions about 100s of project, and only very few maintainers. It doesn’t work if you and I spend equal time on a question. The only way this can work is if you spend a lot of time reading the docs and on composing your questions: what have you tried? What versions are you on? What are you actually trying to do? Why use MDX when you have plain markdown instead of MDX? Otherwise I have to ask these things.

---

> When I wasn't using rehype-raw, the poem (which has HTML in it) rendered empty. Any ideas on that then?

Please reduce the test case, and show the file you mean?

**Reply by hahnbeelee** [reply]
> Why not use MDX?
> What are you suggesting then? I don't follow.

Change your files from `poem.md` and `description.md` to `poem.mdx` and `description.mdx`

**Comment by remcohaszing** [comment]
It appears you’re trying to load both markdown and MDX files. You want to use a rehype plugin for markdown files, but not for MDX. This is possible by specifying two distinct MDX integrations, one for markdown, and one for MDX.

```js
import createMDXPlugin from '@next/mdx'
import rehypeRaw from 'rehype-raw'

const withMDX = createMDXPlugin({
  extension: /\.mdx$/,
  options: {
    // These options apply to .mdx fles only
    providerImportSource: '@mdx-js/react',
    rehypePlugins: [],
    remarkPlugins: [],
  },
})

const withMarkdown = createMDXPlugin({
  extension: /\.md$/,
  options: {
    // These options apply to .md fles only
    providerImportSource: '@mdx-js/react',
    rehypePlugins: [rehypeRaw],
    remarkPlugins: [],
  },
})

const nextConfig = withMDX(withMarkdown({
  compiler: {
    styledComponents: true,
  },
  pageExtensions: ['mdx', 'md', 'tsx', 'ts'],
  poweredByHeader: false,
  // reactStrictMode: true,
  trailingSlash: false,
}))

export default nextConfig
```

**Reply by lancejpollard** [reply]
How can I use `.mdx` (not plain markdown), but still have it allow using native HTML elements? My example of `<li>` above renders blank without the rehype raw plugin. If I can get that rendering then I can just use MDX!

**Reply by lancejpollard** [reply]
I had to do this:

```
const components = {
  Hello: (props: React.ComponentProps<'p'>) => <p {...props} />,
  LI: (props: React.ComponentProps<'li'>) => <li {...props} />,
  OL: (props: React.ComponentProps<'ol'>) => <ol {...props} />,
  P: (props: React.ComponentProps<'p'>) => <p {...props} />,
  UL: (props: React.ComponentProps<'ul'>) => <ul {...props} />,
}
```

Using it like this:

```
<MDXProvider components={components}>{children}</MDXProvider>
```

And havign content like this:

```
<Hello>hey</Hello>

<UL>
  <LI>The moon observes from its perch, high above the peak,</LI>
  <LI>Telling other creatures, their task is never complete.</LI>
  <LI>Far beneath the mountain a valley is carved,</LI>
  <LI>As a river churns for endless years, through stone so hard.</LI>
</UL>
<UL>
  <LI>Approaching like a hunter, the infinite byte</LI>
  <LI>Lighting the way for others to build into the night,</LI>
  <LI>The valley, the breeze, the animals, the trees,</LI>
  <LI>All working together, opening our eyes to see.</LI>
</UL>
<UL>
  <LI>Basking in the warm sun and breathing the mountain air.</LI>
  <LI>Learning from so many, working together is all that's fair.</LI>
  <LI>All that we call nature is constantly changing.</LI>
  <LI>Each day, each hour, every minute rearranging.</LI>
</UL>
<UL>
  <LI>New paths taken as friends come along,</LI>
  <LI>Learning the outcomes in nature are rarely wrong.</LI>
  <LI>There to advise and to learn,</LI>
  <LI>Making sure progress survives every turn.</LI>
</UL>

```

It only shows "hey". I tried with uppercase and lowercase native elements, so something doesn't work with those, any ideas?

**Reply by wooorm** [reply]
It should work without any components, without a provider, with lowercased tag names

**Reply by lancejpollard** [reply]
Hmm, it doesn't appear MDX is using the `components` map I sent to it in the provider... It is rendering `<LI>foo</LI>` when it is at the top level of the MDX file (not nested in another element), but it is showing in the web console as `<p>foo</p>`, even though I said it should be transformed to `<li>`. What am I missing?

**Reply by lancejpollard** [reply]
@wooorm What doesn't it like with my `<ul><li>` example? I removed the rehypeRaw plugin, now it's just not rendering anything when using the list stuff.

**Reply by remcohaszing** [reply]
A couple of things:

- **HTML elements** are case insensitive, but MDX uses **JSX elements**, which are case sensitive.
- If a component starts with an uppercase character, it means it’s a custom component, not a *native* React element.
- This should work though. I tried to reproduce this myself, but it just renders the list as expected for me. I used the latest versions of everything.
- I did manage to reproduce this error by combining [`remark-mdx-frontmatter`](https://github.com/remcohaszing/remark-mdx-frontmatter) and [`rehype-raw`](https://github.com/rehypejs/rehype-raw)
  ```
  error - ./pages/index.mdx
  Error: index.mdx:Error: Cannot compile `mdxjsEsm` node
      at new Promise (<anonymous>)
  ```

**Reply by lancejpollard** [reply]
Thank you for taking a look!

I am using these:

```
"@next/mdx": "^13.1.2",
"@mdx-js/loader": "^2.2.1",
"@mdx-js/mdx": "^2.2.1",
"@mdx-js/react": "^2.2.1",
```

**Reply by wooorm** [reply]
I can only reiterate my last comment: this works, lots of people are using this. You don’t need components, you don’t need providers, an `.mdx` file with an `<ol>` works.
Something else is happening on your end, which might be related to versions, to Next.js, to other configuration you have, etc.
Please make a small reproduction

**Reply by wooorm** [reply]
Why do you have `@mdx-js/mdx`?

**Reply by lancejpollard** [reply]
Not sure.

My minimal example is my fresh blog here https://github.com/lancejpollard/lancejpollard.com, there's not much there yet.

I changed it to use the markdown style list syntax with `<br/>` to separate the lists, and it works for now. So I dunno.

**Reply by wooorm** [reply]
*   That’s not very minimal 😅
*   I don’t know what to do to experience your problems, what should I do?

**Reply by wooorm** [reply]
* There are no `.mdx` files?
* What is `@contents`?

**Reply by wooorm** [reply]
* Looking at dev tools, there clearly are lists and items? You just didn’t style them?

**Reply by wooorm** [reply]
Ok so that’s it, you removed all styles, but it’s working fine.
